src/group/libp2p_group_worker.erl

```diff
@@ -3,253 +3,306 @@
+        _ -> keep_state_and_data
+    end;
+connected(info, {'EXIT', StreamPid, _Reason}, Data=#data{stream_pid=StreamPid}) ->
+    %% The stream we're using died. Let's go back to connecting
+    {next_state, connecting, Data#data{stream_pid=update_stream(undefined, Data)},
+    ?TRIGGER_CONNECT_RETRY};
+connected(cast, clear_target, Data=#data{}) ->
+    %% When the target is cleared we go back to targeting after
+    %% killing the current stream.
+    {next_state, targeting, Data#data{target={undefined, undefined},
+                                      stream_pid=update_stream(undefined, Data)},
+     ?TRIGGER_TARGETING};
+connected(cast, {assign_target, NewTarget}, Data=#data{target=OldTarget}) when NewTarget /= OldTarget ->
+    %% When the target is changed from what we have we kill the
+    %% current stream and go back to targeting.
+    {next_state, targeting, Data#data{target=NewTarget, stream_pid=update_stream(undefined, Data)},
+     ?TRIGGER_TARGETING};
+connected(info, close, Data=#data{}) ->
+    %% On close we just transition to closing. This keeps the stream
+    %% alive for any pending messages until the server decides to
+    %% terminate this worker.
+    {next_state, closing, Data};
 
+connected(EventType, Msg, Data) ->
+    handle_event(EventType, Msg, Data).
 
 
 -spec terminate(Reason :: term(), State :: term(), Data :: term()) -> any().
 terminate(_Reason, _State, Data=#data{connect_pid=Process}) ->
     kill_pid(Process),
-    assign_stream(undefined, Data),
-    monitor_session(undefined, Data).
+    update_stream(undefined, Data).
+
+
+%%
+%% Closing - The worker is trying to shut down. No more connect
+%% attempts are made, but stream assignments are accepted.
+%%
 
+closing(info, {assign_stream, StreamPid}, Data=#data{}) ->
+    %% A new stream assignment came in. If we accept it we stay in
+    %% this state. if not we stay put
+    case handle_assign_stream(StreamPid, Data) of
+        {ok, NewData} -> {keep_state, NewData};
+        _ -> keep_state_and_data
+    end;
+
+closing(EventType, Msg, Data) ->
+    handle_event(EventType, Msg, Data).
 
 
-handle_event(cast, {assign_stream, Conn, StreamPid},
-             Data=#data{stream_pid=_CurrentStreamPid}) when StreamPid /= undefined  ->
+handle_assign_stream(StreamPid, Data=#data{stream_pid=undefined}) ->
+    {ok, update_metadata(Data#data{stream_pid=update_stream(StreamPid, Data)})};
+handle_assign_stream(StreamPid, Data=#data{stream_pid=_CurrentStreamPid}) ->
     %% If send_pid known we have an existing stream. Do not replace.
     case rand:uniform(2) of
         1 ->
             %% lager:debug("Loser stream ~p (addr_info ~p) to assigned stream ~p (addr_info ~p)",
             %%             [StreamPid, libp2p_framed_stream:addr_info(StreamPid),
             %%              _CurrentStreamPid, libp2p_framed_stream:addr_info(_CurrentStreamPid)]),
             libp2p_framed_stream:close(StreamPid),
-            keep_state_and_data;
+            false;
```

Comment:
What does returning false here do?

Reply:
It returns something that isn't the success case? I don't really rely on it, but figured I didn't want to fall over whatever :close returns